### PR TITLE
Implement tag renaming

### PR DESCRIPTION
**Pull request by WinterCore**
### Changelog:
- Fix issue with BatchMutators prepares not being executed before commits
- Change `UserTagsDataLoader` to use a context and move it up the component tree.
- Implement tag renaming.
- Fix issue where subfolders don't get deleted when deleting a folder.

### CODE REVIEW

**Review by burtonator — CHANGES_REQUESTED** [review]
Just a few small changes

File: apps/repository/js/persistence_layer/PersistenceLayerMutator2.tsx [hunk 1]
```diff
@@ -153,142 +162,95 @@ function useRemoveTagsFromDocMetas() {
...
+            tagMutations.push({ ...newTag, id: newTag.label, mutationType: 'add' });
+            userTags.register(newTag);
+        }
 
-                if (! hasTag()) {
-                    // only apply this to items that actually HAVE the tag.
-                    console.log("Skipped annotation without tag: " + docMeta.docInfo.fingerprint, annotation);
-                    return;
-                }
+        // FIXME REVIEW: I was thinking of creating a union type for these 4 ITextHighlight | IAreaHighlight | IFlashcard | IComment
```
Comment (burtonator) [inline apps/repository/js/persistence_layer/PersistenceLayerMutator2.tsx:182]: I'm fine with that... go for it... something like IAnnotationType or something?

File: apps/repository/js/persistence_layer/PersistenceLayerMutator2.tsx [hunk 2]
```diff
@@ -153,142 +162,95 @@ function useRemoveTagsFromDocMetas() {
...
+        const annotationPredicate = (annotation: ITextHighlight | IAreaHighlight | IFlashcard | IComment | IDocAnnotation) => {
+            const tags = (annotation.tags || {});
+            return isPresent(tags[oldTag.id]);
+        };
 
-                const tags = computeNewTags(annotation.tags);
+        // FIXME REVIEW: I was hoping to share some code between prepare & commit but I wasn't able to
```
Comment (burtonator) [inline apps/repository/js/persistence_layer/PersistenceLayerMutator2.tsx:188]: OK... I'll yield to you on this. I don't think it's a problem if refactoring would take a significant amount of time.

File: apps/repository/js/persistence_layer/PersistenceLayerApp.tsx [hunk 3]
```diff
@@ -146,19 +146,19 @@ const UserTagsDataLoaderData = React.memo(function UserTagsDataLoaderData(props:
...
-                            <TagsProviderContext.Provider value={tagsProvider}>
-                            <DocMetaLookupContext.Provider value={docMetaLookupContext}>
-                                {props.children}
-                            </DocMetaLookupContext.Provider>
-                            </TagsProviderContext.Provider>
-                        </TagDescriptorsContext.Provider>
-                    </TagsContext.Provider>
-                </PersistenceLayerContext.Provider>
-            </PersistenceContext.Provider>
+        <PersistenceContext.Provider value={persistenceContext}>
```
[the remark this answers is not in the document]
Comment (WinterCore) [reply apps/repository/js/persistence_layer/PersistenceLayerApp.tsx:148]: Yes

**Review by burtonator — APPROVED** [review]

**Pull request merged**